File: test/integration/test/ResetPasswordSpec.js
'use strict'

const asUser = require('../support/asUser')

const Url = {
  login: '/login',
  resetPassword: (token) => `/reset-password/${token}`,
}

const NewPassword = 'pedEdcitVac8'

class ResetPasswordShortcuts {
  constructor(browser) {
    this.b = browser
  }

  async assertLoggedInAs(email) {
    await this.b.assertExists({ tag: 'div', class: 'logged-in', contains: email, wait: 'pageLoad' })
  }

  async assertLogInFailed() {
    await this.b.assertExists({ tag: 'h1', contains: 'Log in', wait: 'pageLoad' })
    await this.b.assertExists({ class: 'error', contains: 'Wrong email address or password' })
  }

  async assertLoggedOut() {
    await this.b.assertExists({ class: 'session-form', wait: 'pageLoad' })
  }

  async startResetPassword(email) {
    await this.b.click({ link: 'Reset', wait: 'pageLoad' })
    await this.b.sendKeys(email, { css: 'body.password-new input[name=email]', wait: 'pageLoad' })
    await this.b.click('input[value="Email instructions to this address"]')
  }

  async assertShowsResetPasswordForm() {
    await this.b.assertExists('[name=password]', { wait: 'pageLoad' })
    await this.b.assertExists('[name=password2]')
  }

  async fillPasswordsAndClickSubmit(password1, password2) {
    await this.b.sendKeys(password1, '[name=password]')
    await this.b.sendKeys(password2, '[name=password2]')
    await this.b.click('input.btn-primary')
  }

  async tryLogIn(email, password) {
    await this.b.sendKeys(email, '.session-form [name=email]')
    await this.b.sendKeys(password, '.session-form [name=password]')
    await this.b.click('.session-form [type=submit]')
  }

  async logOut() {
    await this.b.click({ link: 'Log out' })
    await this.b.assertExists({ class: 'session-form', wait: 'pageLoad' })
  }
}

// In order to let users store their own information
// We need to let users identify themselves
// Even after they forget their passwords
describe('ResetPassword', function() {
  asUser.usingTemporaryUser(function() {
    before(async function() {
      this.browser.loadShortcuts('jquery')

      this.jquery = this.browser.shortcuts.jquery
      this.resetPassword = new ResetPasswordShortcuts(this.browser)

      this.b = this.browser

      // We'll start this test suite logged out
      await this.resetPassword.logOut()
    })

    describe('When the user tries to reset a password', function() {
      beforeEach(async function() {
        await this.b.get(Url.login)
        await this.resetPassword.startResetPassword(this.userEmail)
      })

      it('should alert the user to check his or her email', async function() {
        await this.b.assertExists({ class: 'alert-success', contains: 'sent', wait: 'pageLoad' })
      })

      it('should do the same thing even on a second attempt', async function() {
        await this.b.get(Url.login)
        await this.resetPassword.startResetPassword(this.userEmail)
        await this.b.assertExists({ class: 'alert-success', contains: 'sent', wait: 'pageLoad' })
      })

      describe('when the user clicks the token from the email', function() {
        beforeEach(async function() {
          const json = await this.adminSession.showUser({ email: this.userEmail })
          const token = JSON.parse(json).reset_password_token
          await this.b.get(Url.resetPassword(token))
          await this.jquery.waitUntilReady()
        })

        it('should show a reset-password form', async function() {
          await this.resetPassword.assertShowsResetPasswordForm()
        })

        it('should not click through if the passwords are not set', async function() {
          await this.resetPassword.fillPasswordsAndClickSubmit('', '')
          await this.resetPassword.assertShowsResetPasswordForm()
        })

        it('should not click through if the passwords do not match', async function() {
          await this.resetPassword.fillPasswordsAndClickSubmit(NewPassword, this.userEmail)
          await this.resetPassword.assertShowsResetPasswordForm()
        })

        it('should not click through if the password is too short', async function() {
          await this.resetPassword.fillPasswordsAndClickSubmit('short', 'short')
          await this.resetPassword.assertShowsResetPasswordForm()
        })

        it('should flash success, log you in and reset the password', async function() {
          await this.resetPassword.fillPasswordsAndClickSubmit(NewPassword, NewPassword)
          await this.b.assertExists({ class: 'alert-success', contains: 'You have updated your password', wait: 'pageLoad' })
          await this.resetPassword.assertLoggedInAs(this.userEmail)
          await this.resetPassword.logOut()

          // old password doesn't work
          await this.resetPassword.tryLogIn(this.userEmail, this.userEmail)
          await this.resetPassword.assertLogInFailed()

          // new password works
          await this.resetPassword.tryLogIn(this.userEmail, NewPassword)
          await this.resetPassword.assertLoggedInAs(this.userEmail)
          await this.resetPassword.logOut()
        })
      })
    })
  })
})

File: integration-test/spec/reset_password_spec.rb
#!/usr/bin/env ruby

require './spec/spec_helper'

# In order to let users store their own information
# We need to let users identify themselves
# Even after they forget their passwords
describe 'Login' do
  before do
    @user = admin_session.create_test_user
    page.visit('/login')
    start_reset_password(@user[:email])
  end

  after do
    admin_session.destroy_test_user(@user)
  end

  def assert_logged_in_as(email)
    page.assert_selector('div.logged-in', text: email, wait: WAIT_LOAD)
  end

  def assert_log_in_failed
    page.assert_selector('h1', text: /Log in/i, wait: WAIT_LOAD)
    page.assert_selector('.error', text: 'Wrong email address or password')
  end

  def assert_logged_out
    page.assert_selector('.session-form', wait: WAIT_LOAD)
  end

  def start_reset_password(email)
    page.click_link('Reset it', wait: WAIT_LOAD) # wait for link to appear
    page.fill_in('Email', with: email, wait: WAIT_LOAD) # wait for page to load
    page.click_button('Email instructions to this address')
  end

  def assert_shows_reset_password_form
    assert page.has_field?('password', wait: WAIT_LOAD) # wait for page to load
    assert page.has_field?('password2')
  end

  def fill_passwords_and_click_submit(password1, password2)
    page.fill_in('password', with: password1, wait: WAIT_LOAD) # wait for page to load
    page.fill_in('password2', with: password2)
    page.click_button('Set new password and log in')
  end

  def try_log_in(email, password)
    page.visit('/login')
    page.within('.session-form') do
      page.fill_in('email', with: email)
      page.fill_in('password', with: password)
      page.click_button('Log in')
    end
  end

  def log_out
    page.click_link('Log out')
  end

  def get_reset_token(user)
    admin_session.get_user_token(user, 'reset-password')
  end

  it 'should alert the user to check his or her email' do
    page.assert_selector('.alert-success', text: 'sent', wait: WAIT_LOAD) # wait for page to load
  end

  it 'should alert even on second attempt' do
    page.visit('/login')
    start_reset_password(@user[:email])
    page.assert_selector('.alert-success', text: 'sent', wait: WAIT_LOAD) # wait for page to load
  end

  describe 'when the user clicks the token from the email' do
    before do
      page.visit("/reset-password/#{get_reset_token(@user)}")
    end

    it 'should not click through if the passwords are not set' do
      fill_passwords_and_click_submit('', '')
      assert_shows_reset_password_form
    end

    it 'should not click through if passwords do not match' do
      fill_passwords_and_click_submit(@user[:password] + 'FOO', @user[:password] + 'BAR')
      assert_shows_reset_password_form
      page.assert_selector('p.help-block', text: 'doesn’t match')
    end

    it 'should not click through if password is too short' do
      fill_passwords_and_click_submit('short', 'short')
      assert_shows_reset_password_form
      page.assert_selector('p.help-block', text: 'characters long')
    end

    it 'should reset the password' do
      new_password = 'gabty3kAg/'
      fill_passwords_and_click_submit(new_password, new_password)
      page.assert_selector('.alert-success', text: 'You have updated your password', wait: WAIT_LOAD) # wait for page to load
      assert_logged_in_as(@user[:email])

      # Old password doesn't work
      log_out
      try_log_in(@user[:email], @user[:password])
      assert_log_in_failed

      # New password works
      try_log_in(@user[:email], new_password)
      assert_logged_in_as(@user[:email])
    end
  end
end
